### PR TITLE
Stop Alluxio processes properly

### PR DESCRIPTION
**Pull request by ggezer**
There are 3 main categories of fixes:
- Register shutdown hooks for all processes
- Make abstract worker implementation restartable in compliance with `Server` interface
- Fix various resource disposal issues

Thse all together will prevent many gRPC level warnings.

### CODE REVIEW

File: job/server/src/main/java/alluxio/worker/JobWorker.java [hunk 1]
```diff
@@ -69,8 +68,7 @@
    * @param ufsManager the ufs manager
    */
   JobWorker(FileSystem filesystem, FileSystemContext fsContext, UfsManager ufsManager) {
-    super(
-        Executors.newFixedThreadPool(1, ThreadFactoryUtils.build("job-worker-heartbeat-%d", true)));
+    super(ExecutorServiceFactories.fixedThreadPool("job-worker-executor", 1));
```
Comment (gpang) [inline job/server/src/main/java/alluxio/worker/JobWorker.java]: is the thread rename intentional?

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 2]
```diff
@@ -147,8 +143,7 @@
   DefaultBlockWorker(BlockMasterClientPool blockMasterClientPool,
       FileSystemMasterClient fileSystemMasterClient, Sessions sessions, BlockStore blockStore,
       UfsManager ufsManager) {
-    super(Executors
-        .newFixedThreadPool(4, ThreadFactoryUtils.build("block-worker-heartbeat-%d", true)));
+    super(ExecutorServiceFactories.fixedThreadPool("block-worker-executor", 4));
```
Comment (gpang) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: is this the heartbeat thread or something else?

File: job/server/src/main/java/alluxio/worker/JobWorker.java (hunk 1 above)
Comment (ggezer) [reply job/server/src/main/java/alluxio/worker/JobWorker.java]: Yes it's actually an worker executor thread. Various heartbeats are running on them.

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java (hunk 2 above)
Comment (ggezer) [reply core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: This is worker executor thread-pool. Various heartbeats are running on them.

File: core/server/worker/src/main/java/alluxio/worker/block/AsyncBlockRemover.java [hunk 3]
```diff
@@ -103,9 +106,12 @@ public void run() {
           mBlockWorker.removeBlock(Sessions.MASTER_COMMAND_SESSION_ID, blockToBeRemoved);
           LOG.debug("Block {} is removed in thread {}.", blockToBeRemoved, mThreadName);
         } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          LOG.warn("{} got interrupted while it was cleaning block {}.",
-              mThreadName, blockToBeRemoved);
+          // Only log warning if interrupted not due to a shutdown.
+          if (!mShutdown) {
+            Thread.currentThread().interrupt();
```
Comment (gpang) [inline core/server/worker/src/main/java/alluxio/worker/block/AsyncBlockRemover.java]: is this supposed to be in the if statement too?

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 4]
```diff
@@ -254,40 +251,35 @@ public void start(WorkerNetAddress address) throws IOException {
...
   @Override
-  public void stop() {
+  public void stop() throws IOException {
     // Steps to shutdown:
     // 1. Gracefully shut down the runnables running in the executors.
-    // 2. Shutdown the executors.
-    // 3. Shutdown the clients. This needs to happen after the executors is shutdown because
+    // 2. Shutdown the clients. This needs to happen after the executors is shutdown because
     //    runnables running in the executors might be using the clients.
+    // 3. Shutdown base worker. (closes executors.)
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: Should we have a mechanism for ensuring the shutdown of resources which need to be closed/stopped(ie. Closer)?

File: core/server/worker/src/main/java/alluxio/worker/block/AsyncBlockRemover.java (hunk 3 above)
Comment (ggezer) [reply core/server/worker/src/main/java/alluxio/worker/block/AsyncBlockRemover.java]: Right. This was incorrect.

Comment (ggezer) [reply core/server/worker/src/main/java/alluxio/worker/block/AsyncBlockRemover.java]: https://github.com/Alluxio/alluxio/pull/10777/commits/275c3d201fb0820e3222cc7e6349f703224b5146

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java (hunk 4 above)
Comment (ggezer) [reply core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: https://github.com/Alluxio/alluxio/pull/10777/commits/41ad003489b98ee5d989f77a4d9bc38cb768c45c

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 5]
```diff
@@ -147,11 +149,11 @@
   DefaultBlockWorker(BlockMasterClientPool blockMasterClientPool,
       FileSystemMasterClient fileSystemMasterClient, Sessions sessions, BlockStore blockStore,
       UfsManager ufsManager) {
-    super(Executors
-        .newFixedThreadPool(4, ThreadFactoryUtils.build("block-worker-heartbeat-%d", true)));
-    mBlockMasterClientPool = blockMasterClientPool;
+    super(ExecutorServiceFactories.fixedThreadPool("block-worker-executor", 5));
+    mResourceCloser = Closer.create();
+    mResourceCloser.register(mBlockMasterClientPool = blockMasterClientPool);
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: `mBlockMasterClientPool = mResourceCloser.register`?

**Comment by calvinjia** [comment]
alluxio-bot, merge this please.

**Pull request merged**